modules/iam/src/main/scala/ch/epfl/bluebrain/nexus/commons/iam/IamClient.scala

```diff
@@ -0,0 +1,96 @@
+package ch.epfl.bluebrain.nexus.commons.iam
+
+import akka.http.scaladsl.client.RequestBuilding.Get
+import akka.http.scaladsl.model.Uri
+import akka.http.scaladsl.model.headers.HttpCredentials
+import cats.syntax.show._
+import cats.{MonadError, Show}
+import ch.epfl.bluebrain.nexus.commons.http.{HttpClient, UnexpectedUnsuccessfulHttpResponse}
+import ch.epfl.bluebrain.nexus.commons.iam.acls.AccessControlList
+import ch.epfl.bluebrain.nexus.commons.iam.auth.User
+import ch.epfl.bluebrain.nexus.commons.iam.identity.Caller
+import ch.epfl.bluebrain.nexus.commons.iam.identity.Caller._
+import ch.epfl.bluebrain.nexus.commons.types.CommonRejections.UnauthorizedAccess
+import journal.Logger
+
+import scala.concurrent.{ExecutionContext, Future}
+
+trait IamClient[F[_]] {
+
+  /**
+    * Retrieve the ''caller'' form the optional [[HttpCredentials]]
+    *
+    * @param optCredentials the optionally provided [[HttpCredentials]]
+    */
+  def getCaller(optCredentials: Option[HttpCredentials]): F[Caller]
+
+  /**
+    * Retrieve the current ''acls'' for some particular ''resource''
+    *
+    * @param resource the resource against which to check the acls
+    * @param caller   the implicitly available [[Caller]]
+    */
+  def getAcls(resource: String)(implicit caller: Caller): F[AccessControlList]
+
+  /**
+    * Retrieve the current ''acls'' for some particular ''resource'' with some ''prefix''
+    *
+    * @param prefix   the prefix the prepend to the resource
+    * @param resource the resource against which to retrieve the acls
+    * @param caller   the implicitly available [[Caller]]
+    */
+  def getAcls[Id](prefix: String, resource: Id)(implicit caller: Caller, S: Show[Id]): F[AccessControlList]
+
+}
+
+object IamClient {
+
+  implicit def akkaHttpIamClient(implicit ec: ExecutionContext,
+                                 aclClient: HttpClient[Future, AccessControlList],
+                                 userClient: HttpClient[Future, User],
+                                 iamUri: IamUri,
+                                 F: MonadError[Future, Throwable]) = new IamClient[Future] {
+    private val log = Logger[this.type]
+
+    override def getCaller(optCredentials: Option[HttpCredentials]) =
+      optCredentials
+        .map { cred =>
+          userClient(getRequest("oauth2/user", optCredentials))
+            .map[Caller](AuthenticatedCaller(cred, _))
+            .recoverWith {
+              case ur: UnexpectedUnsuccessfulHttpResponse =>
```

Comment:
check for 401 and 403 to map to Unauthorized and Forbidden propagate the same error type of others